[HttpFoundation] Allow RedisCluster class for RedisSessionHandler

michaelperrin opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.1 <!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- don't forget to update src/**/CHANGELOG.md files -->
| BC breaks?    | no     <!-- see https://symfony.com/bc -->
| Deprecations? | no <!-- don't forget to update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tests pass?   | yes    <!-- please add some, will be required by reviewers -->
| Fixed tickets |    <!-- #-prefixed issue number(s), if any -->
| License       | MIT
| Doc PR        | 

The `RedisSessionHandler` added in Symfony 4.1 was supposed to accept `RedisCluster` instances but that possibility was not in the condition, leading to an `InvalidArgumentException`.

I formatted the object type condition to allow the `RedisCluster` instance value and also added a test for it.

A double space formatting issue has been fixed too (which is included in the same commit, I can change that if needed).

I could not add a test using RedisCluster, as the installed Redis instance on Travis doesn't use clusters.

michaelperrin committed:
[HttpFoundation] Allow RedisCluster class for RedisSessionHandler

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/RedisSessionHandler.php:
```diff
@@ -33,14 +33,20 @@ class RedisSessionHandler extends AbstractSessionHandler
+            !$redis instanceof \RedisArray &&
+            !$redis instanceof \RedisCluster &&
+            !$redis instanceof \Predis\Client &&
+            !$redis instanceof RedisProxy
```
Missing `\`, same in the docblock

michaelperrin replied:
Do you talk about the RedisProxy class? That one is imported at the top of the class.

nicolas-grekas replied:
Right sorry!

nicolas-grekas reviewed (changes requested):
Thanks

michaelperrin commented on src/Symfony/Component/HttpFoundation/Session/Storage/Handler/RedisSessionHandler.php:
```diff
@@ -33,14 +33,20 @@ class RedisSessionHandler extends AbstractSessionHandler
+        if (
+            !$redis instanceof \Redis &&
+            !$redis instanceof \RedisArray &&
+            !$redis instanceof \RedisCluster &&
```
@nicolas-grekas Just to make things easier to see, this is the condition I added

michaelperrin commented:
Thank you very much @nicolas-grekas, that is indeed useful to have a Redis cluster for tests.

I updated tests in order to use it for tests using `RedisSessionHandler`.

I didn't get why an extra `Redis` instance was created to make checks after values are written, instead of using the same Redis connection instance.
The "validator" instance was always using the `Redis` class connecting to a given host (thus not working when using RedisCluster).

To make everything work, I removed the validator in tests and used instead the `redisClient` instance, which is of the correct type whatever the test.

Maybe you or @dkarlovi will know purpose of the `validator` Redis instance.

Could you make sure that my changes are alright then? If everything looks fine, I will squash commits into one.

dkarlovi commented:
Validator is supposed to be used to connect to the test version of Redis and confirm the intended action happened regardless of the adapter used. Probably shouldn't be changed, but you be the judge, I'm on mobile.

stof commented:
@dkarlovi but shouldn't this `redisClient` be used to connect to redis ? That way, it works for RedisCluster too (otherwise, you would connect to the wrong redis)

michaelperrin commented:
Thanks @dkarlovi for your insight :) I better understand now why you added this other instance.
I would say that it may not be necessary as it is like testing if the Redis libraries work correctly, but I would prefer to be sure.

If we decide it is better to keep the "validator" instance, I will override the other methods in the new `RedisClusterSessionHandlerTest` class to make things work.

dkarlovi commented:
Nice work @michaelperrin! :+1:
